AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/SignIn/SoftwareTokenSetup/CompleteTOTPSetup.swift

```diff
@@ -0,0 +1,141 @@
+
+        do {
+            var deviceMetadata = DeviceMetadata.noData
+            guard let username = signInEventData.username else {
+                throw SignInError.unknown(message: "Unable to unwrap username during TOTP verification")
+            }
+            let authEnv = try environment.authEnvironment()
+            let userpoolEnv = try environment.userPoolEnvironment()
+            let challengeType: CognitoIdentityProviderClientTypes.ChallengeNameType = .mfaSetup
+
+            deviceMetadata = await DeviceMetadataHelper.getDeviceMetadata(
+                for: username,
+                with: environment)
+
+            var challengeResponses = [
+                "USERNAME": username
+            ]
+            let userPoolClientId = userpoolEnv.userPoolConfiguration.clientId
+
+            if let clientSecret = userpoolEnv.userPoolConfiguration.clientSecret {
+                let clientSecretHash = ClientSecretHelper.clientSecretHash(
+                    username: username,
+                    userPoolClientId: userPoolClientId,
+                    clientSecret: clientSecret
+                )
+                challengeResponses["SECRET_HASH"] = clientSecretHash
+            }
+
+            if case .metadata(let data) = deviceMetadata {
+                challengeResponses["DEVICE_KEY"] = data.deviceKey
+            }
+
+            let asfDeviceId = try await CognitoUserPoolASF.asfDeviceID(
+                for: username,
+                credentialStoreClient: authEnv.credentialsClient)
+
+            var userContextData: CognitoIdentityProviderClientTypes.UserContextDataType?
+            if let encodedData = CognitoUserPoolASF.encodedContext(
+                username: username,
+                asfDeviceId: asfDeviceId,
+                asfClient: userpoolEnv.cognitoUserPoolASFFactory(),
+                userPoolConfiguration: userpoolEnv.userPoolConfiguration) {
+                userContextData = .init(encodedData: encodedData)
+            }
+
+            let analyticsMetadata = userpoolEnv
+                .cognitoUserPoolAnalyticsHandlerFactory()
+                .analyticsMetadata()
+
+            let input = RespondToAuthChallengeInput(
+                analyticsMetadata: analyticsMetadata,
+                challengeName: challengeType,
+                challengeResponses: challengeResponses,
+                clientId: userPoolClientId,
+                session: userSession,
+                userContextData: userContextData)
+
+            let responseEvent = try await UserPoolSignInHelper.sendRespondToAuth(
+                request: input,
+                for: username,
+                signInMethod: signInEventData.signInMethod,
+                environment: userpoolEnv)
+            logVerbose("\(#fileID) Sending event \(responseEvent)",
+                       environment: environment)
+            await dispatcher.send(responseEvent)
+            // TODO: HS:
```

Comment:
Remember to remove commented out code before releasing

Reply:
Yep.. This is something I still need to work on.. Will remove it in a follow up PR.